librz/analysis/arch/sh/sh_il.c

```diff
@@ -0,0 +1,1780 @@
+// SPDX-FileCopyrightText: 2022 Dhruv Maroo <dhruvmaru007@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include "sh_il.h"
+#include <rz_il/rz_il_opbuilder_begin.h>
+#include "../../../asm/arch/sh/regs.h"
+
+/**
+ * \file sh_il.c
+ *
+ * Converts SuperH-4 instructions to RzIL statements
+ * References:
+ *  - https://www.st.com/resource/en/user_manual/cd00147165-sh-4-32-bit-cpu-core-architecture-stmicroelectronics.pdf (SH-4 32-bit architecture manual)
+ *  - https://www.renesas.com/in/en/document/mas/sh-4-software-manual?language=en (SH-4 manual by Renesas)
+ *
+ * Both the above references are almost the same
+ *
+ * \note Some things to know before working on this code:
+ * 	- I have used the terms "operand(s)" and "param(s)" interchangeably, and both of them refer to the arguments/params/operands of the instruction
+ *	- `op` doesn NOT mean operand. It is more akin to instruction or opcode. In majority of the places, it means the type of instruction
+ *	- I have used the term "pure" in function names very loosely and shouldn't be used in a literal sense. Refer to the Doxygen documentation to
+ * 	  know what a function does, do not infer it from the function name
+ */
+
+#define SH_U_ADDR(x) UN(SH_ADDR_SIZE, x)
+#define SH_S_ADDR(x) SN(SH_ADDR_SIZE, x)
+#define SH_U_REG(x)  UN(SH_REG_SIZE, (x))
+#define SH_S_REG(x)  SN(SH_REG_SIZE, (x))
+#define SH_BIT(x)    UN(1, x)
+#define SH_TRUE      SH_U_REG(1)
+#define SH_FALSE     SH_U_REG(0)
+
+#define sh_return_val_if_invalid_gpr(x, v) \
+	if (!sh_valid_gpr(x)) { \
+		RZ_LOG_ERROR("RzIL: SuperH: invalid register R%u\n", x); \
+		return v; \
+	}
+
+#define sh_il_get_pure_param(x) \
+	sh_il_get_param(op->param[x], op->scaling).pure
+
+#define sh_il_set_pure_param(x, val) \
+	sh_il_set_param(op->param[x], val, op->scaling)
+
+#define sh_il_get_effective_addr_param(x) \
+	sh_il_get_effective_addr_pc(op->param[x], op->scaling, pc)
+
+/* Utilities */
+
+static bool sh_valid_gpr(ut16 reg) {
+	return reg < SH_GPR_COUNT;
+}
+
+static bool sh_banked_reg(ut16 reg) {
+	return reg < SH_BANKED_REG_COUNT;
+}
+
+/**
+ * \brief Set to true whenever the privilege bit is calculated
+ * It is used to add a `SETL` effect for the privilege bit, in case it is used
+ */
+static bool privilege_check = false;
```

Comment:
Yes, kinda. I can pass a `bool*` to an IL lifter and store the information whether the privilege bit was calculated or not within that `bool`. But this will result in one extra `bool*` in all the instruction lifters and many of the utility functions as well, leading to an ugly interface.

The `privilege_check` is essentially used as a local variable only (since it is initialized to `false` on every invocation of `rz_sh_il_opcode`, but I am using a global variable so as to avoid using an extra argument everywhere.

Reply:
On second thought, I think I should avoid non-const static/global variables since IL lifting has good chances of being multi-threaded in the future. So it's better if I keep that in mind right now as well. I will get rid of the static global.

Reply:
please get rid of this and just check for the register bit